e2e/deployers/applicationset.go

```diff
@@ -10,17 +10,69 @@ import (
 
 type ApplicationSet struct{}
 
-// func (a *ApplicationSet) Init() {
-// }
-
 func (a ApplicationSet) Deploy(w workloads.Workload) error {
-	util.Ctx.Log.Info("enter Deploy " + w.GetName() + "/Appset")
+	name := GetCombinedName(a, w)
+	namespace := util.ArgocdNamespace
 
-	return nil
+	util.Ctx.Log.Info("enter Deploy " + name)
+
+	err := createManagedClusterSetBinding(McsbName, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = createPlacement(name, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = createPlacementDecisionConfigMap(name, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = createApplicationSet(a, w)
+	if err != nil {
+		return err
+	}
+
+	return err
 }
 
 func (a ApplicationSet) Undeploy(w workloads.Workload) error {
-	util.Ctx.Log.Info("enter Undeploy " + w.GetName() + "/Appset")
+	name := GetCombinedName(a, w)
+	namespace := util.ArgocdNamespace
+
+	util.Ctx.Log.Info("enter Undeploy " + name)
+
+	err := deleteApplicationSet(a, w)
+	if err != nil {
+		return err
+	}
+
+	err = deleteConfigMap(name, namespace)
+	if err != nil {
+		return err
+	}
+
+	err = deletePlacement(name, namespace)
+	if err != nil {
+		return err
+	}
+
+	// multiple appsets could use the same mcsb in argocd ns.
+	// so delete mcsb if only 1 appset is in argocd ns
+	lastAppset, err := isLastAppsetInArgocdNs(namespace)
```

Comment:
This can be slightly racy, let's assume that a prallel invocation has progressed to line 24 above, it would have not yet created the AppSet and this check would hence pass and delete the binding.

A simple fix would be to create the binding after creating the AppSet in the Deploy function.

Reply:
this still could not completely remove the race condition.  Thread 1 could know it is last appset, and before it deletes mcsb, thread 2 could create appset and mcsb, and then thread 1 deletes the mcsb.

If we want to fully avoid delete mcsb conflict, a simple way is to just leave mcsb not removed in argocd ns.

Reply:
Also the PlacementDecisionConfigMap is needed by each applicationset. Currently each appset uses different configmap but the cm content is same. 

It should be ok if all appsets use same configmap. But since this need drenv change to create such configmap, I would prefer to use a separate PR to address this.

Reply:
Or, do you want drenv to create both mcsb and PlacementDecisionConfigMap in argocd ns, so that e2e do not need manage them for appset?